File: src/scenes/play/Horde.cpp
#include "Horde.hpp"
#include "Bullet.hpp"
#include <ranges>
#include <CppInvaders.hpp>

#define TIME_STEP 0.016f
#define TIME_FROZEN 1.0f

static const char* sfx[] = {
    SFX_INVADER_STEP0, SFX_INVADER_STEP1,
    SFX_INVADER_STEP2, SFX_INVADER_STEP3};

Horde::Horde() {
    state = DEPLOYING;
    invaders_alive = 0;
    timer = 0;
    i = 0;
}

int Horde::collide_rect(Pico_Abs_Rect rct, Pico_Anchor anc) const {
    Pico_Rel_Rect other_rct = { '!', {(float)rct.x, (float)rct.y, (float)rct.w, (float)rct.h}, anc, NULL };
    for (int i = 0; i < 55; i++) {
        if (!invaders[i].is_alive()) continue;
        Pico_Abs_Rect ir = invaders[i].get_rect();
        Pico_Rel_Rect inv_rct = { '!', {(float)ir.x, (float)ir.y, (float)ir.w, (float)ir.h}, PICO_ANCHOR_N, NULL };
        if (pico_vs_rect_rect(&other_rct, &inv_rct)) {
            return i;
        }
    }

    return -1;
}

int Horde::kill_invader(int i) {
    pico_output_sound(SFX_INVADER_KILLED);
    invaders[i].kill();
    invaders_alive--;

    int score = 10 * (3 - invaders[i].type);
    return score;
}

Bullet* Horde::shoot(float ship_x) {
    Pico_Abs_Pos pos = {0, 0};

    for (const Invader& inv : invaders) {
        bool above_ship = abs(ship_x - inv.x) <= 8;
        if (inv.y > pos.y && above_ship && inv.is_alive()) {
            pos = {inv.x, inv.y};
        }
    }

    if (rand() % 3 && pos.y > 0) {
        return new Bullet((float)pos.x, (float)pos.y + 12, 120);
    }

    std::vector<int> a;
    for (int i = 54; i >= 33; i--) {
        if (invaders[i].is_alive()) {
            a.push_back(i);
        }
    }
    int ri = rand() % a.size(); // FIXME: crashes when horde is empty
    pos = {invaders[a[ri]].x, 0};

    for (const Invader& inv : invaders | std::views::reverse) {
        bool same_column = abs(inv.x - pos.x) <= 8;
        if (inv.y > pos.y && same_column && inv.is_alive()) {
            pos = {inv.x, inv.y};
        }
    }

    return new Bullet((float)pos.x, (float)pos.y + 12, 120);
}

void Horde::update(float delta) {
    Pico_Abs_Dim size;
    pico_get_view(NULL, &size, NULL, NULL, NULL, NULL, NULL, NULL);
    timer += delta;

    switch (state) {
    case DEPLOYING:
        while (timer >= TIME_STEP && i < 55) {
            invaders[i] = Invader(i); i++;
            invaders_alive++;
            timer -= TIME_STEP;
            if (i == 55) {
                state = MARCHING;
                i = 0;
                sfx_i = 0;
                sfx_timer = 0;
                Invader::dx = 2;
                Invader::dy = 0;
                break;
            }
        }
        break;
    case MARCHING:
        for (Invader& inv : invaders) {
            if (inv.state == Invader::DYING) {
                inv.update(delta);
            }
        }

        while (timer >= TIME_STEP) {
            for (; i < 55 && !invaders[i].is_alive(); i++);

            if (i == 55) {
                for (i = 0; i < 55 && !invaders[i].is_alive(); i++);

                Invader::dy = 0;
                for (const Invader& inv : invaders) {
                    bool out = inv.x < 16 || inv.x >= size.w - 16;
                    if (inv.is_alive() && out) {
                        Invader::dx = -Invader::dx;
                        Invader::dy = 8;
                        break;
                    }
                }
            }

            if (i < 55) {
                invaders[i].update(delta); i++;
            }

            timer -= TIME_STEP;
        }

        sfx_timer += delta;
        if (sfx_timer >= TIME_STEP * invaders_alive + 0.1f) {
            pico_output_sound(sfx[sfx_i]);
            sfx_i = (sfx_i + 1) % 4;
            sfx_timer = 0;
        }
        break;
    }
}

void Horde::draw() const {
    for (const Invader& inv : invaders) {
        inv.draw();
    }
}

File: lua/scenes/play/Horde.lua
local class = require "class"
local assets = require "assets"
local Invader = require "scenes.play.Invader"
local Bullet = require "scenes.play.Bullet"

local Horde = class.new()
local sfx_table = { 
    assets.SFX_INVADER_STEP0, 
    assets.SFX_INVADER_STEP1, 
    assets.SFX_INVADER_STEP2, 
    assets.SFX_INVADER_STEP3 
}

function Horde:init()
    self.state = "DEPLOYING"
    self.invaders_alive = 0
    self.timer = 0
    self.i = 0 
    self.sfx_i = 1 
    self.sfx_timer = 0
    self.invaders = {}
end

function Horde:collide_rect(rct, anc)
    local other_rct = { '!', x=rct.x, y=rct.y, w=rct.w, h=rct.h, anchor=anc }
    for i = 0, 54 do
        local inv = self.invaders[i + 1]
        if inv and inv:is_alive() then
            local ir = inv:get_rect()
            local inv_rct = { '!', x=ir.x, y=ir.y, w=ir.w, h=ir.h, anchor='N' }
            if pico.vs.rect_rect(other_rct, inv_rct) then
                return i
            end
        end
    end
    return -1
end

function Horde:kill_invader(i)
    pico.output.sound(assets.SFX_INVADER_KILLED)
    local inv = self.invaders[i + 1]
    inv:kill()
    self.invaders_alive = self.invaders_alive - 1

    local score = 10 * (3 - inv.type)
    return score
end

function Horde:shoot(ship_x)
    local pos = {x=0, y=0}

    for i = 1, 55 do
        local inv = self.invaders[i]
        if inv then
            local above_ship = math.abs(ship_x - inv.x) <= 8
            if inv.y > pos.y and above_ship and inv:is_alive() then
                pos = {x=inv.x, y=inv.y}
            end
        end
    end

    if math.random(0, 2) ~= 0 and pos.y > 0 then
        return Bullet:new(pos.x, pos.y + 12, 120)
    end

    local a = {}
    for i = 54, 33, -1 do
        local inv = self.invaders[i + 1]
        if inv and inv:is_alive() then
            table.insert(a, i)
        end
    end
    
    if #a > 0 then
        local ri = math.random(1, #a)
        local sel_inv = self.invaders[a[ri] + 1]
        pos = {x=sel_inv.x, y=0}

        for i = 55, 1, -1 do
            local inv = self.invaders[i]
            if inv then
                local same_column = math.abs(inv.x - pos.x) <= 8
                if inv.y > pos.y and same_column and inv:is_alive() then
                    pos = {x=inv.x, y=inv.y}
                end
            end
        end
    end

    return Bullet:new(pos.x, pos.y + 12, 120)
end

function Horde:update(delta)
    local view = pico.get.view()
    local size = view.dim
    self.timer = self.timer + delta

    if self.state == "DEPLOYING" then
        while self.timer >= 0.016 and self.i < 55 do
            self.invaders[self.i + 1] = Invader:new(self.i)
            self.i = self.i + 1
            self.invaders_alive = self.invaders_alive + 1
            self.timer = self.timer - 0.016
            if self.i == 55 then
                self.state = "MARCHING"
                self.i = 0
                self.sfx_i = 1
                self.sfx_timer = 0
                Invader.dx = 2
                Invader.dy = 0
                break
            end
        end
    elseif self.state == "MARCHING" then
        for i = 1, 55 do
            local inv = self.invaders[i]
            if inv and inv.state == "DYING" then
                inv:update(delta)
            end
        end

        while self.timer >= 0.016 do
            while self.i < 55 and not self.invaders[self.i + 1]:is_alive() do
                self.i = self.i + 1
            end

            if self.i == 55 then
                self.i = 0
                while self.i < 55 and not self.invaders[self.i + 1]:is_alive() do
                    self.i = self.i + 1
                end

                Invader.dy = 0
                for i = 1, 55 do
                    local inv = self.invaders[i]
                    if inv then
                        local out = inv.x < 16 or inv.x >= size.w - 16
                        if inv:is_alive() and out then
                            Invader.dx = -Invader.dx
                            Invader.dy = 8
                            break
                        end
                    end
                end
            end

            if self.i < 55 then
                self.invaders[self.i + 1]:update(delta)
                self.i = self.i + 1
            end

            self.timer = self.timer - 0.016
        end

        self.sfx_timer = self.sfx_timer + delta
        if self.sfx_timer >= 0.016 * self.invaders_alive + 0.1 then
            pico.output.sound(sfx_table[self.sfx_i])
            self.sfx_i = (self.sfx_i % 4) + 1
            self.sfx_timer = 0
        end
    end
end

function Horde:draw()
    for i = 1, 55 do
        if self.invaders[i] then self.invaders[i]:draw() end
    end
end

return Horde
